src/confd/src/ietf-hardware.c

```diff
@@ -0,0 +1,298 @@
+		free(path);
+		return NULL;
+	}
+	ptr[1] = 0;
+
+	return path;
+}
+
+static int hardware_cand_infer_class(sr_session_ctx_t *session, const char *path)
+{
+	sr_val_t inferred = { .type = SR_STRING_T };
+	struct json_t *root, *usb_ports, *usb_port;
+	sr_error_t err = SR_ERR_OK;
+	json_error_t error;
+	char *name, *class;
+	char *xpath;
+	int index;
+
+	xpath = component_xpath(path);
+	if (!xpath)
+		return SR_ERR_SYS;
+
+	class = srx_get_str(session, "%s/class", xpath);
+	if (class) {
+		free(class);
+		goto out_free_xpath;
+	}
+
+	name = srx_get_str(session, "%s/name", xpath);
+	if (!name) {
+		err = SR_ERR_INTERNAL;
+		goto out_free_xpath;
+	}
+	root = json_load_file("/run/system.json", 0, &error);
+	if (!root) {
+		err = SR_ERR_INTERNAL;
+		goto out_free_name;
+	}
+
+	usb_ports = json_object_get(root, "usb-ports");
+	if (!usb_ports)
+		goto out_free_json; /* No USB-ports is OK */
+
+	json_array_foreach(usb_ports, index, usb_port) {
+		struct json_t *n = json_object_get(usb_port, "name");
+		if (!n || !json_is_string(n)) {
+			ERROR("Did not find hardware port for %s", name);
+			continue;
+		}
+		if (!strcmp(name, json_string_value(n))) {
+			inferred.data.string_val = "infix-hardware:usb";
+			err = srx_set_item(session, &inferred, 0,
+					   "%s/class", xpath);
+			break;
+		}
+	}
+out_free_json:
+	json_decref(root);
+out_free_name:
+	free(name);
+out_free_xpath:
+	free(xpath);
+	return err;
+}
+static int hardware_cand(sr_session_ctx_t *session, uint32_t sub_id, const char *module,
+			 const char *xpath, sr_event_t event, unsigned request_id, void *priv)
+{
+	sr_change_iter_t *iter;
+	sr_change_oper_t op;
+	sr_val_t *old, *new;
+	sr_error_t err = SR_ERR_OK;
+
+	switch (event) {
+	case SR_EV_UPDATE:
+	case SR_EV_CHANGE:
+		break;
+	default:
+		return SR_ERR_OK;
+	}
+	err = sr_dup_changes_iter(session, "/ietf-hardware:hardware/component//*", &iter);
+	if (err)
+		return err;
+	while (sr_get_change_next(session, iter, &op, &old, &new) == SR_ERR_OK) {
+		switch (op) {
+		case SR_OP_CREATED:
+		case SR_OP_MODIFIED:
+			break;
+		default:
+			continue;
+		}
+		err = hardware_cand_infer_class(session, new->xpath);
+		if (err)
+			break;
+	}
+	sr_free_change_iter(iter);
+	return err;
+}
+
+static int change_hardware(sr_session_ctx_t *session, uint32_t sub_id, const char *module,
+			   const char *xpath, sr_event_t event, unsigned request_id, void *priv)
+{
+	struct lyd_node *diff, *difs = NULL, *dif = NULL, *cifs = NULL, *cif = NULL;
+	json_t *root;
+	json_error_t error;
+	sr_data_t *cfg;
+	int rc = SR_ERR_OK;
+
+	switch (event) {
+	case SR_EV_ENABLED: /* first time, on register. */
+	case SR_EV_CHANGE: /* regular change (copy cand running) */
+		break;
+
+	case SR_EV_ABORT: /* User abort, or other plugin failed */
+		return SR_ERR_OK;
+
+	case SR_EV_DONE:
+		/* Check if passed validation in previous event */
+		if (!fexist(SYSFS_CMDS))
+			return SR_ERR_OK;
+
+		if (systemf(SYSFS_CMDS))
+			return SR_ERR_INTERNAL;
+		(void)remove(SYSFS_CMDS);
```

Comment:
If we do decide to keep the current construct with the script file, I think we always want to delete it here - even if `systemf()` fails.